File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/usecases/GiftCardRedeeming.java
package it.giannibombelli.wsc2026.giftcard.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.giftcard.application.commands.RedeemGiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.events.GiftCardRedeemEvents;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;

import static java.util.Objects.requireNonNull;

public final class GiftCardRedeeming implements UseCase<RedeemGiftCard, GiftCardRedeemEvents> {
    private final GiftCardRepository giftCardRepository;

    public GiftCardRedeeming(GiftCardRepository giftCardRepository) {
        this.giftCardRepository = requireNonNull(giftCardRepository);
    }

    @Override
    public GiftCardRedeemEvents invoke(RedeemGiftCard cmd) {
        Require.requireArgument(cmd, "command");
        GiftCardId giftCardId = cmd.aggregateId();

        GiftCard giftCard = giftCardRepository.findById(giftCardId)
            .orElseThrow(() -> new IllegalStateException("GiftCard not found for redeem: " + giftCardId));

        GiftCardRedeemEvents redeemResult = giftCard.redeem(cmd.amount());

        giftCardRepository.save(giftCard);
        return redeemResult;
    }
}

File: typescript/src/giftcard/application/usecases/giftCardRedeeming.ts
// Caso d'uso per il riscatto di una gift card.

import type { UseCase } from '@/common/application/usecase.js';
import type { GiftCardRedeemEvent } from '../../domain/events/giftCardRedeemEvents.js';
import { GiftCardId } from '../../domain/giftcard/giftCardId.js';
import type { GiftCardRepository } from '../../domain/ports/giftCardRepository.js';
import { RedeemGiftCard } from '../commands/redeemGiftCard.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class GiftCardRedeeming implements UseCase<RedeemGiftCard, GiftCardRedeemEvent> {
  private readonly _giftCardRepository: GiftCardRepository;

  constructor(giftCardRepository: GiftCardRepository) {
    requireDependency(giftCardRepository, "giftCardRepository");
    this._giftCardRepository = giftCardRepository;
  }

  invoke(command: RedeemGiftCard): GiftCardRedeemEvent {
    requireArgument(command, 'command');

    const giftCardId: GiftCardId = command.aggregateId;
    const giftCard = this._giftCardRepository.findById(giftCardId);
    if (giftCard === null) {
      throw new Error(`GiftCard not found for redeem: ${giftCardId.value}`);
    }

    const redeemResult = giftCard.redeem(command.amount);

    this._giftCardRepository.save(giftCard);
    return redeemResult;
  }
}
